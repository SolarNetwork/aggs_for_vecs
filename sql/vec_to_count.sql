-- smallint count lots
SELECT vec_to_count(smallints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
-- smallint count none
SELECT vec_to_count(smallints) FROM measurements WHERE sensor_id = -1;
-- smallint count one null
SELECT vec_to_count(smallints) FROM measurements WHERE sensor_id = 1;
-- smallint count array of nulls
SELECT vec_to_count(smallints) FROM measurements WHERE sensor_id = 2;
-- smallint count one not-null
SELECT vec_to_count(smallints) FROM measurements WHERE sensor_id = 4;
-- smallint count array of nulls and one other
SELECT vec_to_count(smallints) FROM measurements WHERE sensor_id IN (2, 4);

-- int count lots
SELECT vec_to_count(ints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
-- int count none
SELECT vec_to_count(ints) FROM measurements WHERE sensor_id = -1;
-- int count one null
SELECT vec_to_count(ints) FROM measurements WHERE sensor_id = 1;
-- int count array of nulls
SELECT vec_to_count(ints) FROM measurements WHERE sensor_id = 2;
-- int count one not-null
SELECT vec_to_count(ints) FROM measurements WHERE sensor_id = 4;
-- int count array of nulls and one other
SELECT vec_to_count(ints) FROM measurements WHERE sensor_id IN (2, 4);

-- bigint count lots
SELECT vec_to_count(bigints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
-- bigint count none
SELECT vec_to_count(bigints) FROM measurements WHERE sensor_id = -1;
-- bigint count one null
SELECT vec_to_count(bigints) FROM measurements WHERE sensor_id = 1;
-- bigint count array of nulls
SELECT vec_to_count(bigints) FROM measurements WHERE sensor_id = 2;
-- bigint count one not-null
SELECT vec_to_count(bigints) FROM measurements WHERE sensor_id = 4;
-- bigint count array of nulls and one other
SELECT vec_to_count(bigints) FROM measurements WHERE sensor_id IN (2, 4);

-- real count lots
SELECT vec_to_count(reals) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
-- real count none
SELECT vec_to_count(reals) FROM measurements WHERE sensor_id = -1;
-- real count one null
SELECT vec_to_count(reals) FROM measurements WHERE sensor_id = 1;
-- real count array of nulls
SELECT vec_to_count(reals) FROM measurements WHERE sensor_id = 2;
-- real count one not-null
SELECT vec_to_count(reals) FROM measurements WHERE sensor_id = 4;
-- real count array of nulls and one other
SELECT vec_to_count(reals) FROM measurements WHERE sensor_id IN (2, 4);

-- float8 count lots
SELECT vec_to_count(floats) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
-- float8 count none
SELECT vec_to_count(floats) FROM measurements WHERE sensor_id = -1;
-- float8 count one null
SELECT vec_to_count(floats) FROM measurements WHERE sensor_id = 1;
-- float8 count array of nulls
SELECT vec_to_count(floats) FROM measurements WHERE sensor_id = 2;
-- float8 count one not-null
SELECT vec_to_count(floats) FROM measurements WHERE sensor_id = 4;
-- float8 count array of nulls and one other
SELECT vec_to_count(floats) FROM measurements WHERE sensor_id IN (2, 4);
