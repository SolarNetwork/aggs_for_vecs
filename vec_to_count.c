
Datum vec_to_count_transfn(PG_FUNCTION_ARGS);
PG_FUNCTION_INFO_V1(vec_to_count_transfn);

/**
 * Returns an of n elements,
 * which each element is the count of non-NULLs found in that position
 * from all input arrays.
 *
 * by Paul A. Jungwirth
 */
Datum
vec_to_count_transfn(PG_FUNCTION_ARGS)
{
  Oid elemTypeId;
  int16 elemTypeWidth;
  bool elemTypeByValue;
  char elemTypeAlignmentCode;
  int currentLength;
  MemoryContext aggContext;
  VecArrayBuildState *state = NULL;
  ArrayType *currentArray;
  int arrayLength;
  Datum *currentVals;
  bool *currentNulls;
  int i;

  if (!AggCheckCallContext(fcinfo, &aggContext)) {
    elog(ERROR, "vec_to_count_transfn called in non-aggregate context");
  }

  // PG_ARGISNULL tests for SQL NULL,
  // but after the first pass we can have a
  // value that is non-SQL-NULL but still is C NULL.
  if (!PG_ARGISNULL(0)) {
    state = (VecArrayBuildState *)PG_GETARG_POINTER(0);
  }

  if (PG_ARGISNULL(1)) {
    // just return the current state unchanged (possibly still NULL)
    PG_RETURN_POINTER(state);
  }
  currentArray = PG_GETARG_ARRAYTYPE_P(1);

  if (state == NULL) {
    // Since we have our first not-null argument
    // we can initialize the state to match its length.
    elemTypeId = ARR_ELEMTYPE(currentArray);
    if (elemTypeId != INT2OID &&
        elemTypeId != INT4OID &&
        elemTypeId != INT8OID &&
        elemTypeId != FLOAT4OID &&
        elemTypeId != FLOAT8OID) {
      ereport(ERROR, (errmsg("vec_to_count input must be array of SMALLINT, INTEGER, BIGINT, REAL, or DOUBLE PRECISION")));
    }
    if (ARR_NDIM(currentArray) != 1) {
      ereport(ERROR, (errmsg("One-dimensional arrays are required")));
    }
    arrayLength = (ARR_DIMS(currentArray))[0];
    // Start with all zeros:
    state = initVecArrayResultWithNulls(elemTypeId, INT8OID, aggContext, arrayLength);
    for (i = 0; i < arrayLength; i++) {
      state->vecvalues[i].i64 = 0;
      state->state.dnulls[i] = false;
    }
  } else {
    elemTypeId = state->inputElementType;
    arrayLength = state->state.nelems;
  }

  get_typlenbyvalalign(elemTypeId, &elemTypeWidth, &elemTypeByValue, &elemTypeAlignmentCode);
  deconstruct_array(currentArray, elemTypeId, elemTypeWidth, elemTypeByValue, elemTypeAlignmentCode,
      &currentVals, &currentNulls, &currentLength);
  if (currentLength != arrayLength) {
    ereport(ERROR, (errmsg("All arrays must be the same length, but we got %d vs %d", currentLength, arrayLength)));
  }

  for (i = 0; i < arrayLength; i++) {
    if (currentNulls[i]) {
      // do nothing: nulls can't change the result.
    } else {
      state->vecvalues[i].i64 += 1;
    }
  }
  PG_RETURN_POINTER(state);
}

Datum vec_to_count_finalfn(PG_FUNCTION_ARGS);
PG_FUNCTION_INFO_V1(vec_to_count_finalfn);

Datum
vec_to_count_finalfn(PG_FUNCTION_ARGS)
{
  Datum result;
  VecArrayBuildState *state;
  int dims[1];
  int lbs[1];
  int i;

  Assert(AggCheckCallContext(fcinfo, NULL));

  state = PG_ARGISNULL(0) ? NULL : (VecArrayBuildState *)PG_GETARG_POINTER(0);

  if (state == NULL)
    PG_RETURN_NULL();

  // Convert from our pgnums to Datums:
  for (i = 0; i < state->state.nelems; i++) {
    state->state.dvalues[i] = Int64GetDatum(state->vecvalues[i].i64);
  }

  dims[0] = state->state.nelems;
  lbs[0] = 1;

  result = makeMdArrayResult(&state->state, 1, dims, lbs, CurrentMemoryContext, false);
  PG_RETURN_DATUM(result);
}
