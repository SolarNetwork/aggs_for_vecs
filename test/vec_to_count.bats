load test_helper

@test "int16 count" {
  result="$(query "SELECT vec_to_count(ints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4)")";
  echo $result;
  [ "$result" = "{2,1,2}" ]
}

# @test "int32 count" {
  # result="$(query "SELECT vec_to_count('{1,1,5,2,0}'::integer[])")";
  # [ "$result" = "0" ]
# }

# @test "int64 count" {
  # result="$(query "SELECT vec_to_count('{1,1,5,2,0}'::bigint[])")";
  # [ "$result" = "0" ]
# }

# @test "float4 count" {
  # result="$(query "SELECT vec_to_count('{1,1,5,2,0}'::real[])")";
  # [ "$result" = "0" ]
# }

@test "float8 count lots" {
  result="$(query "SELECT vec_to_count(floats) FROM measurements WHERE sensor_id IN (1, 2, 3, 4)")";
  echo $result;
  [ "$result" = "{2,1,2}" ]
}

@test "float8 count none" {
  result="$(query "SELECT vec_to_count(floats) FROM measurements WHERE sensor_id = -1")";
  echo $result;
  [ "$result" = "NULL" ]
}

@test "float8 count one null" {
  result="$(query "SELECT vec_to_count(floats) FROM measurements WHERE sensor_id = 1")";
  echo $result;
  [ "$result" = "NULL" ]
}

@test "float8 count array of nulls" {
  result="$(query "SELECT vec_to_count(floats) FROM measurements WHERE sensor_id = 2")";
  echo $result;
  [ "$result" = "{0,0,0}" ]
}

@test "float8 count one not-null" {
  result="$(query "SELECT vec_to_count(floats) FROM measurements WHERE sensor_id = 4")";
  echo $result;
  [ "$result" = "{1,0,1}" ]
}

@test "float8 count array of nulls and one other" {
  result="$(query "SELECT vec_to_count(floats) FROM measurements WHERE sensor_id IN (2, 4)")";
  echo $result;
  [ "$result" = "{1,0,1}" ]
}

@test "string count" {
  run query "SELECT vec_to_count(vals) FROM (VALUES (ARRAY['a']), (ARRAY['b'])) t(vals)"
  echo ${lines}
  [ "${lines[0]}" = "ERROR:  vec_to_count input must be array of SMALLINT, INTEGER, BIGINT, REAL, DOUBLE PRECISION, or NUMERIC" ]
}

@test "numeric count lots" {
  result="$(query "SELECT vec_to_count(nums) FROM measurements WHERE sensor_id IN (1, 2, 3, 4)")";
  echo $result;
  [ "$result" = "{2,1,2}" ]
}

@test "numeric count none" {
  result="$(query "SELECT vec_to_count(nums) FROM measurements WHERE sensor_id = -1")";
  echo $result;
  [ "$result" = "NULL" ]
}

@test "numeric count one null" {
  result="$(query "SELECT vec_to_count(nums) FROM measurements WHERE sensor_id = 1")";
  echo $result;
  [ "$result" = "NULL" ]
}

@test "numeric count array of nulls" {
  result="$(query "SELECT vec_to_count(nums) FROM measurements WHERE sensor_id = 2")";
  echo $result;
  [ "$result" = "{0,0,0}" ]
}

@test "numeric count one not-null" {
  result="$(query "SELECT vec_to_count(nums) FROM measurements WHERE sensor_id = 4")";
  echo $result;
  [ "$result" = "{1,0,1}" ]
}

@test "numeric count array of nulls and one other" {
  result="$(query "SELECT vec_to_count(nums) FROM measurements WHERE sensor_id IN (2, 4)")";
  echo $result;
  [ "$result" = "{1,0,1}" ]
}