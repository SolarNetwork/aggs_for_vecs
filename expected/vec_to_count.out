-- smallint count lots
SELECT vec_to_count(smallints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
 vec_to_count 
--------------
 {2,1,2}
(1 row)

-- smallint count none
SELECT vec_to_count(smallints) FROM measurements WHERE sensor_id = -1;
 vec_to_count 
--------------
 
(1 row)

-- smallint count one null
SELECT vec_to_count(smallints) FROM measurements WHERE sensor_id = 1;
 vec_to_count 
--------------
 
(1 row)

-- smallint count array of nulls
SELECT vec_to_count(smallints) FROM measurements WHERE sensor_id = 2;
 vec_to_count 
--------------
 {0,0,0}
(1 row)

-- smallint count one not-null
SELECT vec_to_count(smallints) FROM measurements WHERE sensor_id = 4;
 vec_to_count 
--------------
 {1,0,1}
(1 row)

-- smallint count array of nulls and one other
SELECT vec_to_count(smallints) FROM measurements WHERE sensor_id IN (2, 4);
 vec_to_count 
--------------
 {1,0,1}
(1 row)

-- int count lots
SELECT vec_to_count(ints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
 vec_to_count 
--------------
 {2,1,2}
(1 row)

-- int count none
SELECT vec_to_count(ints) FROM measurements WHERE sensor_id = -1;
 vec_to_count 
--------------
 
(1 row)

-- int count one null
SELECT vec_to_count(ints) FROM measurements WHERE sensor_id = 1;
 vec_to_count 
--------------
 
(1 row)

-- int count array of nulls
SELECT vec_to_count(ints) FROM measurements WHERE sensor_id = 2;
 vec_to_count 
--------------
 {0,0,0}
(1 row)

-- int count one not-null
SELECT vec_to_count(ints) FROM measurements WHERE sensor_id = 4;
 vec_to_count 
--------------
 {1,0,1}
(1 row)

-- int count array of nulls and one other
SELECT vec_to_count(ints) FROM measurements WHERE sensor_id IN (2, 4);
 vec_to_count 
--------------
 {1,0,1}
(1 row)

-- bigint count lots
SELECT vec_to_count(bigints) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
 vec_to_count 
--------------
 {2,1,2}
(1 row)

-- bigint count none
SELECT vec_to_count(bigints) FROM measurements WHERE sensor_id = -1;
 vec_to_count 
--------------
 
(1 row)

-- bigint count one null
SELECT vec_to_count(bigints) FROM measurements WHERE sensor_id = 1;
 vec_to_count 
--------------
 
(1 row)

-- bigint count array of nulls
SELECT vec_to_count(bigints) FROM measurements WHERE sensor_id = 2;
 vec_to_count 
--------------
 {0,0,0}
(1 row)

-- bigint count one not-null
SELECT vec_to_count(bigints) FROM measurements WHERE sensor_id = 4;
 vec_to_count 
--------------
 {1,0,1}
(1 row)

-- bigint count array of nulls and one other
SELECT vec_to_count(bigints) FROM measurements WHERE sensor_id IN (2, 4);
 vec_to_count 
--------------
 {1,0,1}
(1 row)

-- real count lots
SELECT vec_to_count(reals) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
 vec_to_count 
--------------
 {2,1,2}
(1 row)

-- real count none
SELECT vec_to_count(reals) FROM measurements WHERE sensor_id = -1;
 vec_to_count 
--------------
 
(1 row)

-- real count one null
SELECT vec_to_count(reals) FROM measurements WHERE sensor_id = 1;
 vec_to_count 
--------------
 
(1 row)

-- real count array of nulls
SELECT vec_to_count(reals) FROM measurements WHERE sensor_id = 2;
 vec_to_count 
--------------
 {0,0,0}
(1 row)

-- real count one not-null
SELECT vec_to_count(reals) FROM measurements WHERE sensor_id = 4;
 vec_to_count 
--------------
 {1,0,1}
(1 row)

-- real count array of nulls and one other
SELECT vec_to_count(reals) FROM measurements WHERE sensor_id IN (2, 4);
 vec_to_count 
--------------
 {1,0,1}
(1 row)

-- float8 count lots
SELECT vec_to_count(floats) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
 vec_to_count 
--------------
 {2,1,2}
(1 row)

-- float8 count none
SELECT vec_to_count(floats) FROM measurements WHERE sensor_id = -1;
 vec_to_count 
--------------
 
(1 row)

-- float8 count one null
SELECT vec_to_count(floats) FROM measurements WHERE sensor_id = 1;
 vec_to_count 
--------------
 
(1 row)

-- float8 count array of nulls
SELECT vec_to_count(floats) FROM measurements WHERE sensor_id = 2;
 vec_to_count 
--------------
 {0,0,0}
(1 row)

-- float8 count one not-null
SELECT vec_to_count(floats) FROM measurements WHERE sensor_id = 4;
 vec_to_count 
--------------
 {1,0,1}
(1 row)

-- float8 count array of nulls and one other
SELECT vec_to_count(floats) FROM measurements WHERE sensor_id IN (2, 4);
 vec_to_count 
--------------
 {1,0,1}
(1 row)

-- numeric count lots
SELECT vec_to_count(nums) FROM measurements WHERE sensor_id IN (1, 2, 3, 4);
 vec_to_count 
--------------
 {2,1,2}
(1 row)

-- numeric count none
SELECT vec_to_count(nums) FROM measurements WHERE sensor_id = -1;
 vec_to_count 
--------------
 
(1 row)

-- numeric count one null
SELECT vec_to_count(nums) FROM measurements WHERE sensor_id = 1;
 vec_to_count 
--------------
 
(1 row)

-- numeric count array of nulls
SELECT vec_to_count(nums) FROM measurements WHERE sensor_id = 2;
 vec_to_count 
--------------
 {0,0,0}
(1 row)

-- numeric count one not-null
SELECT vec_to_count(nums) FROM measurements WHERE sensor_id = 4;
 vec_to_count 
--------------
 {1,0,1}
(1 row)

-- numeric count array of nulls and one other
SELECT vec_to_count(nums) FROM measurements WHERE sensor_id IN (2, 4);
 vec_to_count 
--------------
 {1,0,1}
(1 row)

